### ISSUE TITLE
Why doesn't Theano backend for batch_normalization distinguish between train and test modes

### ISSUE THREAD

**Issue by smgutstein** [keras-team/keras#12199]
I'm having trouble getting batch normalization to run at reasonable speeds with a Theano backend. I've noticed that Keras' batch_normalization method does not seem to distinguish between training and testing phases. 

I'm wondering if this is purposeful ? I would have thought that there would be checking to see if a net were in a training or testing phase and then calling T.nnet.bn.batch_normalization_train or T.nnet.bn.batch_normalization_test as required. Is there something I'm not understanding here? Or, would it be reasonable for me to try to make that change?

The code I'm referring to is in theano_backend.py:
```

def batch_normalization(x, mean, var, beta, gamma, axis=-1, epsilon=1e-3):
    """Apply batch normalization on x given mean, var, beta and gamma.
    """
    # TODO remove this if statement when Theano without
    # T.nnet.bn.batch_normalization_test is deprecated
    if not hasattr(T.nnet.bn, 'batch_normalization_test'):
        return _old_batch_normalization(x, mean, var, beta, gamma, epsilon)

    if gamma is None:
        gamma = ones_like(var)
    if beta is None:
        beta = zeros_like(mean)

    if mean.ndim == 1:
        # based on TensorFlow's default: normalize along rightmost dimension
        reduction_axes = list(range(x.ndim - 1))
    else:
        reduction_axes = [i for i in range(x.ndim) if mean.broadcastable[i]]
        
    return T.nnet.bn.batch_normalization_test(
        x, gamma, beta, mean, var, reduction_axes, epsilon)

```
